webhook.go

```diff
@@ -0,0 +1,178 @@
+	binVolumeName = "secrethub-bin"
+
+	// binVolumeMountPath is the mount path where the SecretHub CLI binary can be found.
+	binVolumeMountPath = "/secrethub/bin/"
+)
+
+// binVolume is the shared, in-memory volume where the SecretHub CLI binary lives.
+var binVolume = corev1.Volume{
+	Name: binVolumeName,
+	VolumeSource: corev1.VolumeSource{
+		EmptyDir: &corev1.EmptyDirVolumeSource{
+			Medium: corev1.StorageMediumMemory,
+		},
+	},
+}
+
+// binVolumeMount is the shared volume mount where the SecretHub CLI binary lives.
+var binVolumeMount = corev1.VolumeMount{
+	Name:      binVolumeName,
+	MountPath: binVolumeMountPath,
+	ReadOnly:  true,
+}
+
+// SecretHubMutator is a mutator.
+type SecretHubMutator struct {
+	logger kwhlog.Logger
+}
+
+// Mutate implements MutateFunc and provides the top-level entrypoint for object
+// mutation.
+func (m *SecretHubMutator) Mutate(ctx context.Context, obj metav1.Object) (bool, error) {
+	m.logger.Infof("calling mutate")
+
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		return false, nil
+	}
+
+	version, enabled := pod.Annotations["secrethub"]
+	if !enabled {
+		m.logger.Debugf("Skipping pod %s because it is not annotated with secrethub", pod.Name)
+		return false, nil
+	}
+
+	mutated := false
+
+	for i, c := range pod.Spec.InitContainers {
+		c, didMutate, err := m.mutateContainer(ctx, &c)
+		if err != nil {
+			return false, err
+		}
+		if didMutate {
+			mutated = true
+			pod.Spec.InitContainers[i] = *c
+		}
+	}
+
+	for i, c := range pod.Spec.Containers {
+		c, didMutate, err := m.mutateContainer(ctx, &c)
+		if err != nil {
+			return false, err
+		}
+		if didMutate {
+			mutated = true
+			pod.Spec.Containers[i] = *c
+		}
+	}
+
+	// binInitContainer is the container that pulls the SecretHub CLI
+	// into a shared volume mount.
+	var binInitContainer = corev1.Container{
+		Name:            "copy-secrethub-bin",
+		Image:           "secrethub/cli" + ":" + version,
+		ImagePullPolicy: corev1.PullIfNotPresent,
+		Command: []string{"sh", "-c",
+			fmt.Sprintf("cp /usr/bin/secrethub %s", binVolumeMountPath)},
+		VolumeMounts: []corev1.VolumeMount{
+			{
+				Name:      binVolumeName,
+				MountPath: binVolumeMountPath,
+			},
+		},
+	}
+
+	// If any of the containers requested SecretHub secrets, mount the shared volume
+	// and ensure the SecretHub CLI is available via an init container.
+	if mutated {
+		pod.Spec.Volumes = append(pod.Spec.Volumes, binVolume)
+		pod.Spec.InitContainers = append([]corev1.Container{binInitContainer},
+			pod.Spec.InitContainers...)
+	}
+
+	return false, nil
+}
+
+// mutateContainer mutates the given container, updating the volume mounts and
+// command if it contains SecretHub references.
+func (m *SecretHubMutator) mutateContainer(_ context.Context, c *corev1.Container) (*corev1.Container, bool, error) {
+	// Ignore if there are no SecretHub references in the container.
+	if !m.hasSecretHubReferences(c.Env) {
+		return c, false, nil
+	}
```

Comment:
What if a container has a `secrethub.env` file (maybe provided by a volume)?

Reply:
Interesting use case. We've considered the use of `secrethub.env` files in Kubernetes before, but figured that in that case you'd include the `secrethub.env` file in the image and therefore it would make sense to include the CLI in the image as well. I have not yet considered including the `secrethub.env` file via a volume. How do you imagine this would work? Where would the `secrethub.env` file live? What are the advantages of having the `secrethub.env` file over configuring the environment in the kube files?

Reply:
Well. The main thing I am hinting at: do we want to do this automatic behaviour if someone has explicitly requested injection of the CLI through an annotation?

Reply:
I'm not sure.

Currently, you're annotating a pod to indicate that SecretHub should be attached to a container when secret references are used for that container. When a pod contains multiple containers, you may not want to include SecretHub for every container.

We could move the annotation to container level. This gives the user more control at the cost of having to annotate multiple times if multiple containers within the pod use SecretHub.

Reply:
Hmm, it appears you cannot annotate containers.

Reply:
I'm only now seeing @florisvdg suggestion here: https://github.com/secrethub/secrethub-kubernetes-mutating-webhook/pull/1#discussion_r435876446

That would solve annotating on container level.